Doc to describe how to use JSON type definitions (#716)
brimsec/zq: zeek/README.md

The file [`types.json`](types.json) in this directory contains configuration
that can be used to restore Zeek-compatible rich data types to events that
have been output in JSON format by Zeek. The definitions in the current
revision of `types.json` reflect the default set of logs output by
[Zeek v3.1.2](https://github.com/zeek/zeek/releases/tag/v3.1.2) and are
likely to be compatible with default logs from older Zeek/Bro releases as
well.

If your Zeek environment is customized, you will need to adjust the type
definition to match the differences in your JSON logs. The sections below
describe the customization process.

# Contact us!

If you're using JSON type defintions with `zq` or [Brim](https://github.com/brimsec/brim),
we'd like to hear from you! We know the process of customizing the definitions
can be tricky. We have ideas for ways we might improve it, but we'll have a
better sense of priority and how to go about it if we've heard from those
who have tried it. Whether you've used this feature and it "just worked" or if
you hit challenges and need help, please join our
[public Slack](https://join.slack.com/t/brimsec/shared_invite/zt-cy34xoxg-hZiTKUT~1KdGjlaBIuUUdg)

> and tell us about it. Thanks!

Reviewer: Shouldn't we also mention GitHub issues here for bug reports?

Author: Sure. Added.

# Sample data

To see a working example before working with your own data, clone the
[zq-sample-data](https://github.com/brimsec/zq-sample-data) repository.

```
# git clone --depth=1 https://github.com/brimsec/zq-sample-data ~/zq-sample-data
```

A set of JSON-format Zeek logs will now be present in `~/zq-sample-data/zeek-ndjson`.

# Usage

> Asssuming you have the [`zq`](https://github.com/brimsec/zq) repository locally

Reviewer: Should we briefly mention the existence of a json type inference mechanism, just so readers have a sense of what happens if you omit `-j` altogether?

Author: I added some stuff to mention inference in a section low down. Let me know what you think.

cloned to `~/zq` and the `zq` binary is in your `$PATH`, here's an example of
reading in all Zeek events while applying the JSON type definition:

```

> # zq -f table -j ~/zq/zeek/types.json "count()" ~/zq-sample-data/zeek-ndjson/*

Reviewer: In this case, the `count()` would be the same whether or not we had the typing mechanism (which can be verified by omitting `-j` to fall back on plain json type inference). How about a query that requires zeek types ? A simple example that comes to mind would be `every :x: count`. Or maybe something involving IPs/subnets.

Reviewer: Hmm, now I see that the queries below _do_ rely on types, so this comment is probably moot.

Author: My goal here was just to do something that would cause it to read all events and generate errors. I thought about doing `*` and redirecting the output to `/dev/null` but went this route instead.

COUNT
1462078
```

Since we saw no errors from `zq`, that means the definitons in `types.json`
fully described all named fields in all the log files. If there had been
errors, `zq` would have stopped reading when the first error was encountered
and would not have returned a `count()` result. The sections below describe
what to do if we'd seen such errors here.

# Why is this even necessary?

Consider this Zeek HTTP event as output by the
[JSON Streaming Logs](https://github.com/corelight/json-streaming-logs)
package:

```

> # gzcat ~/zq-sample-data/zeek-ndjson/http.ndjson.gz | head -1 | jq -S .

Reviewer: nit: isn't `head -n 1` more usual?

Author: Hunh. I use `head` so rarely that I'd never noticed that. I just did some reading and saw that it's shorthand that's undocumented in many implementations. So, yeah... changing it to `-n 1`.

{
  "_path": "http",
  "_write_ts": "2018-03-24T17:15:20.610930Z",
  "host": "10.47.3.200",
  "id.orig_h": "10.164.94.120",
  "id.orig_p": 36729,
  "id.resp_h": "10.47.3.200",
  "id.resp_p": 80,
  "method": "GET",
  "request_body_len": 0,
  "resp_fuids": [
    "FnHkIl1kylqZ3O9xhg"
  ],
  "resp_mime_types": [
    "text/html"
  ],
  "response_body_len": 56,
  "status_code": 301,
  "status_msg": "Moved Permanently",
  "tags": [],
  "trans_depth": 1,
  "ts": "2018-03-24T17:15:20.609736Z",
  "uid": "CpQfkTi8xytq87HW2",
  "uri": "/chassis/config/GeneralChassisConfig.html",
  "user_agent": "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)",
  "version": "1.1"
}
```

> To appreicate the importance of correct data typing, it helps to compare this

Reviewer: ```suggestion
To appreciate the importance of correct data typing, it helps to compare this
```

absent from the default Zeek v3.1.2 output that the type definition is based
on.

Now that we understand the problem, we could manually add the missing
configuration to our `types.json`. But this would be tedious and error-prone.
To make this easier, a Zeek script
[`print-types.zeek`](https://github.com/brimsec/zeek/blob/master/brim/print-types.zeek)
exists that will output a complete type definition for each event type that
may be generated by Zeek. By running it in our customized Zeek installation,
we'll get a customized type definition that's unique for our environment.

Following the comments in the script, we run it while ensuring to invoke the
`local` configuration so that all customizations are included:
```
# ZEEK_ALLOW_INIT_ERRORS=1 zeek print-types.zeek local | grep descriptors | jq -S . > types-custom.json
warning in /usr/local/zeek-3.1.2/share/zeek/site/local.zeek, line 106: Loading script '__load__.bro' with legacy extension, support for '.bro' will be removed in Zeek v4.1
Skipping openflow log as it has records nested within records
```

(Note: The "Skipping openflow..." message is due to known issue [#15](https://github.com/brimsec/zeek/issues/15)

> that should be soon addressed. It does not affect this example.)

Reviewer: ```suggestion
It does not affect this example. If you have logs with multiple levels of nesting, please let us know and/or comment on the issue.)
```

>       {
>         "name": "sshka",
>         "type": "bstring"
>       },
>       {
>         "name": "hasshServerAlgorithms",
>         "type": "bstring"
>       },
>       {
3394a3423,3444
>     "unknown_mime_type_discovery_log": [
>       {
>         "name": "_path",
>         "type": "string"
>       },
>       {
>         "name": "ts",
>         "type": "time"
>       },
>       {
>         "name": "fid",
>         "type": "bstring"
>       },
>       {
>         "name": "bof",
>         "type": "bstring"
>       },
>       {
>         "name": "_write_ts",
>         "type": "time"
>       }
>     ],
3797a3848,3852
>       "descriptor": "unknown_mime_type_discovery_log",
>       "name": "_path",
>       "value": "unknown_mime_type_discovery"
>     },
>     {
```

Re-reading our data with the new custom definitions, we no longer see errors.
The higher event count reflects events that are no longer rejected by `zq`
due to the type definitions that were previously absent.

```
# zq -f table -e=false -j types-custom.json "count()" json_streaming_*.log
COUNT
1463414
```

# Older versions of Zeek/Bro

Since the `types.json` provided here is based on Zeek v3.1.2, you may wonder

> if it's appropriate to use it if you're running an older verison of Zeek.

Reviewer: ```suggestion
if it's appropriate to use it if you're running an older version of Zeek.
```

Cut fields ts,id.orig_h not present together in input
```

> Because we didn't apply a typing definition, here `zq` was performing _type_

Reviewer: ```suggestion
Because we didn't apply a typing definition (via the `-j` flag), here `zq` was performing _type_
```
